mitodl/mit-open
# Refactor course numbers

# What are the relevant tickets?
Closes #138

# Description (What does it do?)
- Adds a `course_numbers` `JSONField` to the `Course` model, to be populated in the format:
  ```
  {
    "value": "12.01",
    "listing_type": <"Primary" or "Cross-linked">,
    "department: {
        "department_id": "12",
        "name": "Earth, Atmospheric and Planetary Sciences"
    }
  }
  ```
- Removes the `Course.extra_course_numbers` field.
- Adds migrations to update the schema and convert existing courses to have populated values for `course_numbers`
- Updates the serializers and ETL pipelines as necessary to support the above.
- adds additional fields that are needed for indexing, in this case, modifying `course.course_numbers` to include ``sorted_coursenum` and `primary`.
- Modified the search api to exclude from search results any additional fields defined above, via `SOURCE_EXCLUDED_FIELDS` constant (which will need to be updated as necessary).


# How can this be tested?
- Run `docker compose up`.  The migrations should run, and afterward, if you have any existing courses, the API should return results with each course resource having a populated `{"course": {"course_numbers"}}` section.
- Run `./manage.py recreate_index --all`
- All the ETL pipelines should still work without errors.
- Import an OCW course with extra course numbers if you don't already have any:
  ```
  ./manage.py backpopulate_ocw_data --course-name 4-401-environmental-technologies-in-buildings-fall-2018
  ```
- Search for that course by both the primary and extra course number, it should be returned in both searches:
  - http://localhost:8063/api/v1/learning_resources_search/?q=4.401
  - http://localhost:8063/api/v1/learning_resources_search/?q=4.464
- You should not see "sorted_coursenum" or "primary" fields in the `course.course_numbers` response
- The primary value for this course should be "4.401" and not "4.401+fall_2018" like the `readable_id`
-  Modify [this code](https://github.com/mitodl/mit-open/pull/173/files#diff-fc36e9091d5012053128f4d7ec939c6a644f38629fb4afa9d1f47f7a7a8f770cR342-R344) so that `SOURCE_EXCLUDED_FIELDS=[]`
- Try the above queries again, results should be the same but now `sorted_coursenum` and `primary` fields should be shown.
